tensorflow_probability/examples/stochastic_block_model.py

```diff
@@ -0,0 +1,86 @@
+import time
+
+from matplotlib import pylab as plt
+import numpy as np
+from observations import karate
+import tensorflow as tf
+from tensorflow_probability import edward2 as ed
+
+
+# Data & parameters
+# -----------------------------------------------------------------------------
+x_data, z = karate('~/data')
+V = x_data.shape[1]
+K = 2
+learning_rate = 1e-4
+max_steps = 10000
+epsilon= 0.001
+
+
+def compute_loss(latent_vars, data):
+    """ Compute the loss associated with MAP calculations.
+    """
+    dict_vals = {var: latent.value for var, latent in latent_vars.items()}
+    for x, datum in data.items(): # assuming we have the values as input
+        dict_vals[x] = datum
+    
+    log_prob = 0.0
+    for z in latent_vars.keys():
+        z_copy = z.distribution.copy()
+        log_prob += tf.reduce_sum(z_copy.log_prob(dict_vals[z]))
```

Comment:
I think this indirection is probably necessary because youre using Edward's wrapped TF distribution. Is there some advantage of using Edward here vs directly using TF distributions?

I ask because the typical use case of Edward would be to call make_log_joint from what is otherwise the generative specification. In your case you dont need make_log_joint so I think this example could me made simpler by directly using TF distributions, eg, similar to the Bayesian GMM colab.

Thoughts?

Reply:
Hi! Thanks for your help! The only advantage I see is simplicity: the model specification is a literal translation of the equations; I find it more explicit and expressive. It would make sense for me to implement something like `find_MAP` in edward2.

Now, I understand why it would make more sense here (and now) to use TF distributions directly. I will work on that. In the meantime, do you have any idea why the current scripts does not converge?

Reply:
I'd actually to push back on Josh's comment. You're writing a log joint which is just the reduce sum of individual distribution's log probs. This is clearly an application of Edward2 as the direct translation as a generative program + `ed.make_log_joint` is clearer.

Maybe to be more convincing, can you show generations or predictions of held out elements? This requires more than a log-joint and without Edward2, would require writing down the model twice.